File: algorithms/24. Swap Nodes in Pairs/solution.cc
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        ListNode dummy(0);
        dummy.next = head;
        auto cur = & dummy;
        // if there are more than two nodes left
        while ( cur && cur->next && cur->next->next ) {
            // record the first of current pair and the first of next pair
            auto t1 = cur->next;
            auto t2 = cur->next->next->next;
            // swap
            cur->next = cur->next->next;
            cur->next->next = t1;
            // move to next pair
            t1->next = t2;
            cur = t1;
        }
        return dummy.next;
    }
};

File: algorithms/24. Swap Nodes in Pairs/solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def swapPairs(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        cur = dummy = ListNode(0)
        dummy.next = head
        while cur and cur.next and cur.next.next:
            cur.next.next.next, cur.next.next, cur.next = cur.next, cur.next.next.next, cur.next.next
            cur = cur.next.next
        return dummy.next
